### PR TITLE
[libcxx] renames some template type parameters

### PR DESCRIPTION
**Pull request by cjdb**
Sweeps through the libc++ headers to rename the following template type paramters:

* Iterators are renamed to `_Iter` unless the category is ambiguous. When it is ambiguous:
  * Input iterators are renamed to `_InIter`
  * Output iterators are renamed to `_OutIter`
* Sentinels are renamed to `_Sent`
* General invocable objects are renamed to `_Func`

### CODE REVIEW

**Comment by hawkinsw** [comment]
See also https://github.com/llvm/llvm-project/pull/76201

File: libcxx/include/future [hunk 1]
```diff
@@ -844,7 +844,7 @@ void __async_assoc_state<_Rp, _Func>::__execute() {
 #endif // _LIBCPP_HAS_NO_EXCEPTIONS
 }
 
-template <class _Rp, class w_Func>
+template <class _Rp, class _Func>
```
Comment (mordante) [inline libcxx/include/future:847]: Interesting this worked at all :-/ likewise for the other.

File: libcxx/docs/Contributing.rst [hunk 2]
```diff
@@ -45,6 +45,22 @@ other implementations (e.g. system headers), the test in
 ``libcxx/test/libcxx/system_reserved_names.gen.py`` contains the list of
 reserved names that can't be used.
 
+We use the following names to refer to various tempalate type parameters:
```
Comment (mordante) [inline libcxx/docs/Contributing.rst:48]: Thanks a lot for adding documentation!

File: libcxx/docs/Contributing.rst [hunk 3]
```diff
@@ -45,6 +45,22 @@ other implementations (e.g. system headers), the test in
...
+  * ``_BiIter`` to disambiguate input iterators
+  * ``_RandomIter`` to disambiguate input iterators
+  * ``_ContgiuousIter`` to disambiguate input iterators
+  * ``_OutIter`` to disambiguate output iterators
+* ``_Sent``, to refer to sentinels
+* ``_Func``, to refer to arbitrary invocable objects
+* ``_Pred``, to refer to predicates
+* ``_Comp``, to refer to comparators
+* ``_Proj``, to refer to projections
+
```
Comment (mordante) [inline libcxx/docs/Contributing.rst:63]: Maybe add a comment that older code may use other conventions. For example in `format` I use `_OutIt`

File: libcxx/docs/Contributing.rst [hunk 4]
```diff
@@ -45,6 +45,22 @@ other implementations (e.g. system headers), the test in
...
+  * ``_ForwardIter`` to disambiguate input iterators
+  * ``_BiIter`` to disambiguate input iterators
+  * ``_RandomIter`` to disambiguate input iterators
+  * ``_ContgiuousIter`` to disambiguate input iterators
+  * ``_OutIter`` to disambiguate output iterators
+* ``_Sent``, to refer to sentinels
+* ``_Func``, to refer to arbitrary invocable objects
+* ``_Pred``, to refer to predicates
+* ``_Comp``, to refer to comparators
+* ``_Proj``, to refer to projections
```
Comment (mordante) [inline libcxx/docs/Contributing.rst:62]: How about adding the common `_Tp` and `_Up`?

File: libcxx/test/std/thread/thread.threads/thread.thread.class/thread.thread.constr/constr.compile.fail.cpp [hunk 5]
```diff
@@ -9,12 +9,11 @@
 // <thread>
 
 // class thread
-//     template <class _Fp, class ..._Args,
-//         explicit thread(_Fp&& __f, _Args&&... __args);
+//     template <class _Func, class ..._Args,
+//         explicit thread(_Func&& __f, _Args&&... __args);
```
Comment (mordante) [inline libcxx/test/std/thread/thread.threads/thread.thread.class/thread.thread.constr/constr.compile.fail.cpp:13]: Actually here we should use the non-ugly names as used in the Standard's synopsis.

**Review by mordante — COMMENTED** [review]
Thanks a lot for the cleanup! Some minor nits.

File: libcxx/docs/Contributing.rst (hunk 2 above)
Comment (hawkinsw) [inline libcxx/docs/Contributing.rst:48]:
```suggestion
We use the following names to refer to various template type parameters:
```

File: libcxx/docs/Contributing.rst [hunk 6]
```diff
@@ -45,6 +45,22 @@ other implementations (e.g. system headers), the test in
...
 
+We use the following names to refer to various tempalate type parameters:
+
+* ``_Iter``, to refer to an iterator type. Some alterations are used if multiple
+  iterators with different categories appear in the same template:
+  * ``_InIter`` to disambiguate input iterators
+  * ``_ForwardIter`` to disambiguate input iterators
+  * ``_BiIter`` to disambiguate input iterators
+  * ``_RandomIter`` to disambiguate input iterators
+  * ``_ContgiuousIter`` to disambiguate input iterators
```
Comment (hawkinsw) [inline libcxx/docs/Contributing.rst:56]:
```suggestion
  * ``_ContiguousIter`` to disambiguate input iterators
```

**Review by hawkinsw — COMMENTED** [review]
Hope that these fixups help!

**Comment by hawkinsw** [comment]
> See also #76201

Should we attempt to merge these at roughly the same time, @mordante ?

File: libcxx/docs/Contributing.rst [hunk 7]
```diff
@@ -45,6 +45,22 @@ other implementations (e.g. system headers), the test in
 ``libcxx/test/libcxx/system_reserved_names.gen.py`` contains the list of
 reserved names that can't be used.
```
Comment (ldionne) [inline libcxx/docs/Contributing.rst:46]: I am fine with the change itself, but I question the added value of the documentation. Indeed, this naming scheme works well for e.g. algorithms where the iterators (or predicates, or whatever) don't have any semantics associated to them. However, in other parts of the code, it may convey more information to call the iterator parameter something other than its iterator category. I think it would be better to let people make the right naming decision based on what their code does. However, in cases where we're e.g. writing a new algorithm, folks would end up starting from (or getting inspiration from) code that consistently has the naming we want, so this might end up not being an issue at all.

Basically, I have a slight concern that documenting this naming scheme is going to be too strict, since there are cases where this naming scheme is not what we want. But either way I'm fine with this.

Maybe rewording to

> We **generally** use the following names to refer to various template type parameters:

would alleviate my concern.

**Review by ldionne — COMMENTED** [review]
LGTM with a comment about the documentation.

File: libcxx/test/std/thread/thread.threads/thread.thread.class/thread.thread.constr/constr.compile.fail.cpp (hunk 5 above)
Comment (var-const) [reply libcxx/test/std/thread/thread.threads/thread.thread.class/thread.thread.constr/constr.compile.fail.cpp:13]: +1 -- it's a preexisting issue, but I think we previously agreed that when copying from synopsis, we should just do a direct copy-paste without adjusting any names or formatting. I'm ok with the change (IMO fixing it is out of scope for this patch) or with simply not touching this file.

File: libcxx/docs/Contributing.rst (hunk 7 above)
Comment (var-const) [reply libcxx/docs/Contributing.rst:46]: +1 -- the new documentation is great, but IMO should make it clear that these names are used when more specific names are not available, i.e. it's not a 100% requirement to use this naming scheme. I think adding "generally" would pretty much solve this, but I could also suggest adding a note like:
> This naming scheme should be used for consistency in all cases where the types don't have any semantics associated with them and thus using a more specific name is not possible.

**Review by var-const — COMMENTED** [review]
IMO this is awesome, thanks a lot for doing this! Almost LGTM except the comment re. documentation.

**Comment by ldionne** [comment]
Is there still interest for pursuing this? If not, I would close to clear up the review queue.
